### DISCUSSION TITLE
Load external data inside custom validator during validation

### DISCUSSION THREAD

**Discussion by duke4** [networknt/json-schema-validator#1047] [Q&A] [answered]
Hi there,

I took a look into the Docs and existing discussions, but didn't find an answer to my question. So I hope someone can help me with that:
I wonder if it is possible to load external data inside a custom validator during validation. Loading external data means to query a database or something similar. Are collectors a way to go here? If yes, I'm not sure how. I don't really get it from the Docs.

Currently my implementation looks similar to this code:

```java
JsonMetaSchema customMetaSchema = JsonMetaSchema.builder(JsonMetaSchema.getV202012())
  .keyword(new CustomKeyword())
  .build()

JsonSchemaFactory jsonSchemaFactory = new JsonSchemaFactory.Builder().defaultMetaSchemaIri(customMetaSchema.getIri())
  .metaSchema(customMetaSchema)
  .build();

JsonSchema jsonSchema = jsonSchemaFactory.getSchema(jsonNode);

ValidationResult validationResult = jsonSchema.validateAndCollect(node);
```

And the Collector would be something like this:

```java
public class CustomCollector implements Collector<List<String>> {

    private final ExternalService externalService;

    public CustomCollector(ExternalService externalService) {

        this.externalService = externalService;
    }


    @Override
    public void combine(Object object) {

    }

    @Override
    public List<String> collect() {

        return externalService.loadExternalData();
    }
}
```

If the collector is the right choice to do this, what is the way of including them into the validation and use the collected data inside a custom validator?
Or is the collector only to collect results from a validator to be able to access it afterwards? What would then be correct way for my use case?

TIA

**Comment by justin-tay** [comment]
Not clear what your use case is. What are you going to do with this data?

If you want to load data in a validator you would just have a service in your custom validator implementation to load the data. Nothing to do with a collector.

**Reply by duke4** [reply]
You're right, sorry. It's not getting that clear.

Yes I can have the service in the validator, but the "problem" is the following:
Currently the meta schema and the factory are created only once and re-used for every validation.
But the single validations can be triggered by different users and the user data is also needed inside the validator to call the service.

Is there a smart way of getting this into the validation execution when creating the schema and validate against it?
Or is it necessary to create the meta schema every time a new validation is triggered?

**Reply by justin-tay** [reply, marked as answer]
You can pass data on a per execution basis through the `ExecutionContext`.

```java
Set<ValidationMessage> messages = schema.validate(inputData, InputFormat.JSON, executionContext -> {
    executionContext.getCollectorContext().getCollectorMap().put("userData", "myUserData");
});
```

The `validate` method in the validator has a `executionContext` parameter which you can use to get the data.

**Reply by duke4** [reply]
Alright perfect. That seems to work as intended 👍 
Thank you
